exchanges/coinbasepro/coinbasepro_test.go

```diff
@@ -2,679 +2,1453 @@ package coinbasepro
 	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	status = make([]string, 0)
+	assets = make([]string, 1)
+	assets[0] = testCrypto.String()
+	_, err = c.GetAllOrders(context.Background(), "", "USD", "LIMIT", "SELL", "", "SPOT", "RETAIL_ADVANCED",
+		"UNKNOWN_CONTRACT_EXPIRY_TYPE", "2", status, assets, 10, time.Time{}, time.Time{})
+	assert.NoError(t, err)
+}
 
-	_, err := c.GetAccounts(context.Background())
-	if err != nil {
-		t.Error("GetAccounts() error", err)
-	}
-	accountResponse, err := c.GetAccount(context.Background(),
-		"13371337-1337-1337-1337-133713371337")
-	if accountResponse.ID != "" {
-		t.Error("Expecting no data returned")
-	}
-	if err == nil {
-		t.Error("Expecting error")
-	}
-	accountHistoryResponse, err := c.GetAccountHistory(context.Background(),
-		"13371337-1337-1337-1337-133713371337")
-	if len(accountHistoryResponse) > 0 {
-		t.Error("Expecting no data returned")
-	}
-	if err == nil {
-		t.Error("Expecting error")
+func TestGetFills(t *testing.T) {
+	_, err := c.GetFills(context.Background(), "", "", "", time.Unix(2, 2), time.Unix(1, 1), 0)
+	assert.ErrorIs(t, err, common.ErrStartAfterEnd)
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	_, err = c.GetFills(context.Background(), "", testPair.String(), "", time.Unix(1, 1), time.Now(), 5)
+	assert.NoError(t, err)
+	status := []string{"OPEN"}
+	ordID, err := c.GetAllOrders(context.Background(), "", "", "", "", "", "", "", "", "", status, nil, 3, time.Time{},
+		time.Time{})
+	assert.NoError(t, err)
+	if ordID == nil || len(ordID.Orders) == 0 {
+		t.Skip(skipInsufficientOrders)
+	}
+	_, err = c.GetFills(context.Background(), ordID.Orders[0].OrderID, "", "", time.Time{}, time.Time{}, 5)
+	assert.NoError(t, err)
+}
+
+func TestGetOrderByID(t *testing.T) {
+	_, err := c.GetOrderByID(context.Background(), "", "", "")
+	assert.ErrorIs(t, err, errOrderIDEmpty)
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	ordID, err := c.GetAllOrders(context.Background(), "", "", "", "", "", "", "", "", "", nil, nil, 10,
+		time.Time{}, time.Time{})
+	assert.NoError(t, err)
+	if ordID == nil || len(ordID.Orders) == 0 {
+		t.Skip(skipInsufficientOrders)
+	}
+	resp, err := c.GetOrderByID(context.Background(), ordID.Orders[0].OrderID, ordID.Orders[0].ClientOID,
+		testFiat.String())
+	assert.NoError(t, err)
+	assert.NotEmpty(t, resp, errExpectedNonEmpty)
+}
+
+func TestPreviewOrder(t *testing.T) {
+	_, err := c.PreviewOrder(context.Background(), "", "", "", "", "", 0, 0, 0, 0, 0, 0, false, false, false, time.Time{})
+	assert.ErrorIs(t, err, errAmountEmpty)
+	_, err = c.PreviewOrder(context.Background(), "", "", "", "", "", 0, 1, 0, 0, 0, 0, false, false, false, time.Time{})
+	assert.ErrorIs(t, err, errInvalidOrderType)
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	skipTestIfLowOnFunds(t)
+	resp, err := c.PreviewOrder(context.Background(), testPair.String(), "BUY", "MARKET", "", "", 0, 1, 0, 0, 0, 0, false,
+		false, false, time.Time{})
+	assert.NoError(t, err)
```

Comment:
I am receiving the following:
`CoinbasePro unsuccessful HTTP status code: 400 raw response: {"error":"unknown","error_details":"proto: (line 1:78): invalid value for enum type: \"\"","message":"proto: (line 1:78): invalid value for enum type: \"\""}, authenticated request failed`
This seems to suggest an invalid request. I'm receiving this with multiple authorised requests. Please let me know if you are passing these tests and I'll look into what's going on for me a bit further before commenting

Reply:
Ah, that slipped under my testing. The marginType parameter was added later; when I went back to add it I neglected to remove skipTestIfLowOnFunds. Since I didn't have enough funds to properly run it, that had me never check if margin_type could be sent as an empty string, and my assumption that it could was wrong.

I've fixed that oversight. Now, I fail with the error "account is not available", which I suspect means that I have insufficient funds. Further info would be appreciated!